Brick `__brick_copy` specialized by `_ExecutionPolicy` without `std::decay_t`

SergeyKopienko opened the pull request:
In this PR we fix the issue https://github.com/uxlfoundation/oneDPL/issues/2110

mmichel11 commented:
How about just removing `_ExecutionPolicy` from the class templates of `__brick_copy` and `__brick_copy_n`? If we take a look at the our specializations of `__brick_copy`, first on the host:
https://github.com/uxlfoundation/oneDPL/blob/417f118fb2363908962e3871bf7ce970c32ca264/include/oneapi/dpl/pstl/algorithm_impl.h#L1079-L1105 and secondly in the hetero backend:
https://github.com/uxlfoundation/oneDPL/blob/417f118fb2363908962e3871bf7ce970c32ca264/include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h#L416-L425 We see we use the tag to define these specializations whereas `_ExecutionPolicy` is used.

Maybe I am misunderstanding something, but to me it looks like all we need is a tag.

SergeyKopienko commented:
I don't know the history why `_ExecutionPolicy` is used as brick's template parameter, so right now I don't think that their remove is correct. May be someone else explain these reasons.

SergeyKopienko commented:
> > I don't know the history why `_ExecutionPolicy` is used as brick's template parameter, so right now I don't think that their remove is correct. May be someone else explain these reasons.
> 
> My guess is that when we added tag dispatching, we replaced the needed functionality from the policy with functionality from the Tag, but neglected to remove the policy parameter then. I agree that we can just remove this template parameter.

Yes, @danhoeflinger, you are right:
 - as example I found the previous state of
 ```C++
template <typename _ExecutionPolicy>
struct __brick_copy<_ExecutionPolicy,
                    oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy, void>>
{
    // ...
}
```

MikeDvorskiy commented:
> ```c++
> oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy, void>>
> ```

Yes, before the tag dispatching policy type is used for type specialization..
But, now, when we have type specialization based on tag type trait,  we don't need `_ExecutionPolicy` parameter.

SergeyKopienko commented:
> > ```c++
> > oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy, void>>
> > ```
> 
> Yes, before the tag dispatching policy type is used for type specialization.. But, now, when we have type specialization based on tag type trait, we don't need `_ExecutionPolicy` parameter.

As far as I understood @rarutyun on our offline discussion with him, he has opposite opinion....

SergeyKopienko commented:
Continue in the PR https://github.com/uxlfoundation/oneDPL/pull/2129
